File: backend/app/models/novel.py
from datetime import datetime
from sqlalchemy import Boolean, Column, String, Integer, DateTime, ForeignKey
from sqlalchemy.orm import relationship

from ..database import Base


class Novel(Base):
    __tablename__ = "novels"

    id = Column(String(36), primary_key=True, index=True)
    title = Column(String(255), nullable=False)
    description = Column(String(500), nullable=True)
    genre = Column(String(100), nullable=True)
    status = Column(String(50), default="draft")
    cover_image = Column(String(500), nullable=True)
    tags = Column(String(255), nullable=True)
    is_banned = Column(Boolean, default=False, nullable=False)
    word_count = Column(Integer, default=0)
    chapter_count = Column(Integer, default=0)
    user_id = Column(String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    user = relationship("User", back_populates="novels")
    chapters = relationship("Chapter", back_populates="novel", cascade="all, delete")
    outlines = relationship("Outline", back_populates="novel", cascade="all, delete")
    characters = relationship("Character", back_populates="novel", cascade="all, delete")
    world_building = relationship("WorldBuilding", back_populates="novel", cascade="all, delete", uselist=False)

File: frontend/src/services/novel.ts
import api from "./api";

export type Novel = {
  id: string;
  title: string;
  description?: string;
  status: string;
  genre?: string;
  tags?: string;
  chapter_count?: number;
  word_count?: number;
};

export async function listNovels(): Promise<Novel[]> {
  const { data } = await api.get<Novel[]>("/novels/");
  return data;
}

export async function createNovel(payload: { title: string; description?: string }): Promise<Novel> {
  const { data } = await api.post<Novel>("/novels/", payload);
  return data;
}

export async function getNovel(novelId: string): Promise<Novel> {
  const { data } = await api.get<Novel>(`/novels/${novelId}`);
  return data;
}

export async function updateNovel(
  novelId: string,
  payload: Partial<Pick<Novel, "title" | "description" | "status" | "genre" | "tags">>
): Promise<Novel> {
  const { data } = await api.put<Novel>(`/novels/${novelId}`, payload);
  return data;
}

export async function deleteNovel(novelId: string): Promise<void> {
  await api.delete(`/novels/${novelId}`);
}
